# Exploring eBay Car Sales Data
This dataset is of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. This data was originally scraped and uploaded to Kaggle by user but is no longer available. Now this dataset is hosted on Dataquest and a few modifications from the original dataset:
- 50,000 data points were sampled from the full dataset
- the dataset was made to more closely resemble what you would typically expect from a scraped dataset.

The data dictionary is provided with the data fields below:
- **dateCrawled** - when this ad was first crawled. All field values were taken from this
- **name** - name of the car
- **seller** - where the seller is private or a dealer
- **offerType** - the type of listing
- **price** - the price on the ad to sell the car
- **abtest** - whether the listing is included in an A/B test
- **vehicleType** - the vehicle type
- **yearOfRegistration** - the year in which the car was first registered
- **gearbox** - the transmission type
- **powerPS** - the power of the car in PS
- **model** - the car model name
- **kilometer** - how many kilometers the car has driven
- **monthOfRegistration** - the month in which the car was first registered
- **fuelType** - what type of fuel the car uses
- **brand** - the brand of the car
- **notRepairedDamage** - if the car has a damage which is not yet repaired
- **dateCreated** - the date on which the eBay listing was created
- **nrOfPictures** - the number of pictures in the ad
- **postalCode** - the postal code for the location of the vehicle
- **lastSeenOnline** - when the crawler saw this ad last one

# Project Goal
The goal of this project is to clean the data and analyze the included used car listings. It's also increases the familiarity of the unique benefits jupyter notebook provides for pandas.

# Getting Started
To start I want to import the libraries we need and reading the dataset into pandas

In [48]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [49]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

Now I want to use the DataFrame.info() method to print information about the **autos** dataframe

In [50]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## Initial Observations
From the info() method, what I first notice is that a few of the categories have null values, like **fuelType**, **vehicleType**, and three other categories. These will need to be cleaned before I analyze them. There are 20 columns and 5000 entries. The entries are broken down into two dtypes: int64 and strings, mostly strings. The column names use camelcase instead of Python's preferred snakecase and that means I can't just replace spaces with underscores

In [51]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Now I want to convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

I am going to use the DataFrame.columns attribute to print an array of the exisiting column names

In [52]:
new_array = autos.columns
print(new_array)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


Now I want to copy that array and make the following edits to columns names:
- yearOfRegistration to registration_year
- monthOfRegistration to registration_month
- notRepairedDamage to unrepaired_damage
- dateCreated to ad_created
And assign the modified column names back to the **DataFrame.columns** attribute. To check my current state of the autos dataframe and see the changes to the dataframe, I will use **DataFrame.head()**

In [53]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head()



date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Basic Data Exploration
Now I want to look through the dataset to determine what other cleaning tasks need to be done. Initially I'm looking for:
- text columns where all or almost all values are the same, because these can often be dropped as they aren't typically useful for analysis
- examples of numeric data stored as text which can be cleaned and converted

First I will use **DataFrame.describe()** with the include=ALL to look at descriptive statistics for all columns, categorical and numeric columns.


In [54]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

So far I'm observing that the information in several of the values have null unique values. Columns like registration_year, power_ps, registration_month, num_photos and postal_code all have NaN for unique and top values. These columns could need additional investigation as to why that is. 

There are also several categories with only 2 unique values so almost all of the 5000 entries in these categories are the same. We can see that in seller, offer_type, unrepaired_damage and more.

I do see a few examples of numeric data stored as text in the price and odometer columns. So for each column, I will:
- remove any non-numeric characters
- convert the column to a numeric dtype
- then use DataFrame.rename() to rename the column to odometer_km

In [55]:
print(autos['price'])
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','').astype(int)
print(autos['price'])

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: object
0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64


In [56]:
print(autos['odometer'])
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',', '').astype(int)
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)
print(autos['odometer_km'])

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object
0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int64


## Additional observations - price and odometer
Now that I have done some basic cleaning on the price and odometer_km columns by converting them to numeric types and removing non-numeric values. Now I want to analyze these columns using minimum and maximum values and look for any values that look unrealistically high or low which are outliers that we might want to remove.

To do this I'm going to use the methods outlined below:

- Series.unique().shape to see how many unique values
- Series.describe() to view min/max/median/mean
- Series.value_counts() with some variations:
    - chained to .head() if there are lots of values
    - because Series.value_counts() returns a series, we can use Series.sort_index() wth ascending=True or False to view the highest and lowest values with their counts
- remove outliers using df[df["col"].between(x,y)]

Starting with the price column.

In [57]:
prices = autos['price']
print(prices)
unique_prices = prices.unique().shape
price_counts = prices.value_counts().head()
highest = price_counts.sort_index(ascending=True)
lowest = price_counts.sort_index(ascending=False)
print(unique_prices)
print(price_counts)
prices.describe()

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64
(2357,)
0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

From these observations, I've noticed that standard deviation is very high. This means that there are some outliers that we need to address. First I want to consider values higher than 50,000 and apply Series.values_counts() to check the number of times these values occur.

In [58]:
high_price = autos[autos['price'] >= 50000]
high_price['price'].value_counts().sort_values(ascending=False)

52900     5
55000     5
56000     4
53000     4
60000     3
         ..
58900     1
71000     1
72900     1
116000    1
82987     1
Name: price, Length: 152, dtype: int64

Above 50,000, the total number of cars for sale are 152 and only one price appears multiple times, (52,900 - 5 times). These values look unrealistically high, so I am going to ignore the values about 50,000. Now lets check values that are lower than $100.

In [59]:
low_price = autos[autos['price'] < 100]
low_price['price'].value_counts().head()

0     1421
1      156
50      49
99      19
80      15
Name: price, dtype: int64

I can see that there are 1421 values that equal to 100. These values could easily be discarded. I think for future analysis that I should only consider the values between 100 and 50,000. 

In [60]:
new_price = autos[autos['price'].between(100,50000)]
new_price['price'].describe()

count    48038.000000
mean      5611.942691
std       6723.459745
min        100.000000
25%       1250.000000
50%       3000.000000
75%       7400.000000
max      50000.000000
Name: price, dtype: float64

After removing the outliers on both ends of the data, the standard deviation is much lower than before indicating that the values are more uniform

In [61]:
miles = autos['odometer_km']
miles.unique().shape
miles.describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Odometer's column statistical details show that the values are almost uniformly distributed as the standard deviation(std) is low. So I do not see an adequate reason to remove any lower boundary values because there will not be much difference among those values

# Converting date columns
Moving on to the date columns and understand the date range the data covers.

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:
- 'date_crawled': added by the crawler
- 'last_seen': added by the crawler
- 'ad_created': from the website
- 'registration_month': from the website
- 'registation_year': from the website

Currently date_crawled, last_seen, ad_created columns are all identified as string values by pandas. They need to be converted into a numerical representation to understand it quantitatively. Before I can do that I need to understand how the values are formatted.

- First I want to extract just the date values
  -  select the first 10 characters in each column using Series.str[:10]
- use series.value_counts() to generate a distribution
  - to include missing values in the distribution and to use percentages instead of counts, chain the Series.value_counts(normalize=True,dropna=False)
- sort by the index
  - rank by date in ascending order


In [62]:
#Apply Series.split() method and extract the values of first position in the date_created column
autos['date_crawled'] = autos['date_crawled'].str.split(r' ').str[0]

#Apply Series.value_counts()  method on the row
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

While applying Series.value_counts() method we have included missing values and percentages instead of counts in the distribution. We have also used Series.sort_index() in order to rank the date in ascending order (earliest date to latest date).

From the above analysis we make following observations. date_crawled column does not have any null values. All the data are crawled between March 2016 and April 2016. The highest number of data is crawled on 3rd April 2016 (~ 4%) and lowest on 7th April 2016 (~0.1%).

Below we perform the same analysis on the date_created column.

In [63]:
autos['ad_created'] = autos['ad_created'].str.split(r' ').str[0]
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

The ad_created column also does not have any null values. The listings in this dataset were created between June 2015 and April 2016. The highest number of listing was done on April 3rd, 2016 (~4%). Lastly I want to perform the same analysis on the last seen column.

In [64]:
autos['last_seen'] = autos['last_seen'].str.split(r' ').str[0]
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

From the above analysis we make following observations. last_seen column has no null values. Here most of the ad's are last seen by the crawler between the month of March and April of 2016. Most of the ad's are seen on 6th April 2016 (~ 2.2%) and least are on 5th March 2016 (~ 0.01%).

Further we study the registration_year column by applying Series.describe() method on it.

In [65]:
years = autos['registration_year']
years.describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## Findings from Registration_Year
One of the first things that I notice in the years is a max of 9999 and a minimum value of 1000. Seeing as how the year 1000 was long before the advent of the car and the year 9999 is long into the future.  
The rest of them seem to be between 1990s and 2010s. Especially since the data was pulled in 2016. Any date above 2016 is definitely inaccurate. However determining the earliest valid year is more difficult. So let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely or if I need more custom logic.

In [66]:
bf_1900 = autos[autos['registration_year'] < 1900]
bf_1900['registration_year']

10556    1800
22316    1000
24511    1111
32585    1800
35238    1500
49283    1001
Name: registration_year, dtype: int64

In [67]:
af_2016 = autos[autos['registration_year'] > 2016]
af_2016['registration_year']

10       2017
55       2017
65       2017
68       2017
84       2018
         ... 
49796    2017
49841    2017
49880    2017
49910    9000
49935    2017
Name: registration_year, Length: 1966, dtype: int64

From the above analysis, I can see that there are only six cars registered before 1900 and there are a total of 1966 cars which are registered above year 2016. However because there are no records in this data set listing ads after 2016. So It's best to ignore those points which are less than 4% of the total data. The lowest year should be in the early 1900s or late 1800s since that is when the gas powered engine car was invented. So Its best to set the lower limit at 1900.

In [68]:
new_ry = autos[autos['registration_year'].between(1900,2016)]

# look into the stats of corrected column
new_ry['registration_year'].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

From the new corrected column, I can see that the mean year of registration is 2002. Now lets apply the value_counts() method on the edited column

In [69]:
new_ry['registration_year'].value_counts(normalize=True)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

The normalized value_counts show that most of the registrations ar done in the year 2000, around 7% and the least amount of registrations were done in 1952, less than 1%.

## Exploring Price by Brand
When working with data on cars, its import to explore variations across different car brands. So in this section, I will use aggregation to understand the brand column. This process looks like:
- identify the unique values to aggregate by
- create an empty dictionary to store aggregate data
- loop over the unique values and for each:
    - create a subset for the unique values
    - calculate the mean 
    - asign the val/mean to the dictionary as a key value pair
    
Let's start with listing the unique names of the top 20 car brands.

In [70]:
unique_brand = autos['brand'].value_counts().index[0:20]
unique_brand

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')

In [71]:
# check number of each brand
autos['brand'].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

From this information, I can tell that it would be best to use the first 20 unique brands as they contain a higher number of vehicles. This is evident by the top brand volkswagen having nearly double the number of vehicles as the next highest.

So, now I want to construct a dictionary with unique brand name as key and mean value of price as value

In [72]:
unique_brands_dict = {}
for key in unique_brand:
    selected_brand = new_price[new_price['brand'] == key]
    unique_brands_dict[key] = selected_brand['price'].mean()

unique_brands_dict

{'volkswagen': 5354.53715064758,
 'opel': 2974.764503159104,
 'bmw': 8076.7559762860965,
 'mercedes_benz': 8130.28503046127,
 'audi': 8955.433939393939,
 'ford': 3631.9298507462686,
 'renault': 2411.617787418655,
 'peugeot': 3086.930281690141,
 'fiat': 2815.635782747604,
 'seat': 4348.652792990142,
 'skoda': 6394.309677419355,
 'mazda': 4075.319293478261,
 'nissan': 4681.94046008119,
 'smart': 3538.344927536232,
 'citroen': 3783.6788856304984,
 'toyota': 5148.0032733224225,
 'sonstige_autos': 9637.480725623584,
 'hyundai': 5416.23382045929,
 'volvo': 4911.680459770115,
 'mini': 10566.824940047962}

Now I am going to sort the average price in the above dictionary, so that I can easily check which brands have a higher average price and which have a lower average price.

In [73]:
#swap the key and value in the above dictionary and sort the values
unique_brands_swap = dict([(value, key) for key, value in unique_brands_dict.items()])
values = unique_brands_swap.items()
sorted_values = sorted(values, reverse=True)
sorted_values

[(10566.824940047962, 'mini'),
 (9637.480725623584, 'sonstige_autos'),
 (8955.433939393939, 'audi'),
 (8130.28503046127, 'mercedes_benz'),
 (8076.7559762860965, 'bmw'),
 (6394.309677419355, 'skoda'),
 (5416.23382045929, 'hyundai'),
 (5354.53715064758, 'volkswagen'),
 (5148.0032733224225, 'toyota'),
 (4911.680459770115, 'volvo'),
 (4681.94046008119, 'nissan'),
 (4348.652792990142, 'seat'),
 (4075.319293478261, 'mazda'),
 (3783.6788856304984, 'citroen'),
 (3631.9298507462686, 'ford'),
 (3538.344927536232, 'smart'),
 (3086.930281690141, 'peugeot'),
 (2974.764503159104, 'opel'),
 (2815.635782747604, 'fiat'),
 (2411.617787418655, 'renault')]

The average price for the selected brands ranges from around 10000 to 2000. Brand mini has highest average price i.e. ~10566. Next four brands with highest average price are sonstige_autos, audi, mercedes_benz and bmw. At the bottom of the list there are smart, peugeot, opel, fiat and the least average price is for renault i.e. ~2411.

## Storing Aggregate Data in a DataFrame

First I want to look at the top 6 brands in our top 20 unique brands list.

In [74]:
top_brands = unique_brand[:6]
top_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

If we look at the average prices of these 6 brands, we observe that:-

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between

For these brands we would like to check if there is a link between average mileage and mean price. Below we will construct dictionaries for mean price and mean mileage for these 6 brands.

In [75]:
mean_price = {}
mean_mileage = {}
for key in top_brands:
    selected_top_brands = new_price[new_price['brand'] == key]
    mean_price[key] = selected_top_brands['price'].mean()
    mean_mileage[key] = selected_top_brands['odometer_km'].mean()
mean_price

{'volkswagen': 5354.53715064758,
 'opel': 2974.764503159104,
 'bmw': 8076.7559762860965,
 'mercedes_benz': 8130.28503046127,
 'audi': 8955.433939393939,
 'ford': 3631.9298507462686}

In [76]:
mean_mileage

{'volkswagen': 129031.55127081508,
 'opel': 129442.84893739231,
 'bmw': 133076.11397972843,
 'mercedes_benz': 131554.61270670147,
 'audi': 130124.84848484848,
 'ford': 124489.55223880598}

It is very difficult to compare these two dictionaries. I need to construct a pandas series for both these and join them using pandas dataframe constructor. This will enable an easy comparision between the two columns (mean_price and mean_mileage).

In [79]:
mean_price_series = pd.Series(data=mean_price)
mean_price_series

volkswagen       5354.537151
opel             2974.764503
bmw              8076.755976
mercedes_benz    8130.285030
audi             8955.433939
ford             3631.929851
dtype: float64

In [80]:
mean_mileage_series = pd.Series(data=mean_mileage)
mean_mileage_series

volkswagen       129031.551271
opel             129442.848937
bmw              133076.113980
mercedes_benz    131554.612707
audi             130124.848485
ford             124489.552239
dtype: float64

In [81]:
#construct a dataframe for mean price
top_brand_data = pd.DataFrame(data=mean_price_series, columns=['price_mean'])
top_brand_data

price_mean
volkswagen     5354.537151
opel           2974.764503
bmw            8076.755976
mercedes_benz  8130.285030
audi           8955.433939
ford           3631.929851

In [82]:
#add mean mileage series to the abovem dataframe
top_brand_data['mileage_mean']= mean_mileage_series[:]
top_brand_data.astype('int64') #convert the data-type to int-64

price_mean  mileage_mean
volkswagen           5354        129031
opel                 2974        129442
bmw                  8076        133076
mercedes_benz        8130        131554
audi                 8955        130124
ford                 3631        124489

The mean mileage of all the top 6 brands are in the range125000 km - 133000 km. Highest mean mileage is for bmw, which is ~ 133076 km and the lowest is for ford, which is ~ 124489 km. The mean price is lowest for opel i.e. ~2974 and highest for audi at ~8955. However I do not observe any pattern for the relationship between mean price and mean mileage.

# Conclusions
In conclusion, the project goal was to work with a data set containing the details of used cars from eBay Klenanzeigen, the classified section of German eBay. The main objectives were Data Cleaning and Exploration and Data Analysis.
- **Data Cleaning** - Starting with cleaning the column names from camelcase to snakecase, as well as changing the names of columns for better clarity. I explored both the price and odometer column by examining the statistics of column-values and dropping the unrealistic and inaccurate values. Next I explored the date columns and dropped all incorrect registration year column values.
- **Data Analysis** - I performed two main analysis on this data set
    - Exploring Price by Brand: Here I selected the top 20 brands and calculated the average of thse brands. The average price ranged from 2000-10000. With the brand mini having the highest average price at 10,566. The next four brands with the highest average price were sonstige autos, audi, mercedes benz and bmw. The least average price was for the brand renault at 2,411.
    - Mean Price & MRan Mileage of Top Six Brands: For this one, I calculated the mean mileage and mean price for six unique brands (volkswagen, opel, bmw, mercedes benz and ford). The mean mileage of all these brands are in the range 125,000 km to 133,000km. The highest mean mileage was for bmw at 133,076 km and the lowest was ford at 124,489 km. The mean price was lowest for opel at 2974 and highest for audi at 8955.

# Additional Next Steps
There are several ways to go after the initial cleaning and analysis. Here are a few ideas for expanding this project.
- Data Cleaning next steps:
    - identify categorical data that uses german words, translate them and map the values to their english counterparts
    - convert the dates to be uniform numeric data, from "2016-03-21" to 20160321
    - see if there are particular keywords in the name column to be extracted as new columns
- Analysis next steps:
    - Find the most common brand/model combinations
    - Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
    - How much cheaper are cars with damage than their non-damaged counterparts?